kiwi/filesystem/__init__.py

```diff
@@ -15,24 +15,18 @@
 # You should have received a copy of the GNU General Public License
 # along with kiwi.  If not, see <http://www.gnu.org/licenses/>
 #
-# project
-from kiwi.filesystem.ext2 import FileSystemExt2
-from kiwi.filesystem.ext3 import FileSystemExt3
-from kiwi.filesystem.ext4 import FileSystemExt4
-from kiwi.filesystem.btrfs import FileSystemBtrfs
-from kiwi.filesystem.xfs import FileSystemXfs
-from kiwi.filesystem.fat16 import FileSystemFat16
-from kiwi.filesystem.fat32 import FileSystemFat32
-from kiwi.filesystem.squashfs import FileSystemSquashFs
-from kiwi.filesystem.clicfs import FileSystemClicFs
-from kiwi.filesystem.swap import FileSystemSwap
-
-from kiwi.exceptions import (
-    KiwiFileSystemSetupError
+import importlib
+from typing import Dict
+from abc import (
+    ABCMeta,
+    abstractmethod
 )
 
+# project
+from kiwi.exceptions import KiwiFileSystemSetupError
+
 
-class FileSystem:
+class FileSystem(metaclass=ABCMeta):
```

Comment:
That makes FileSystem an abstract base class. It's a way of duck-typing that allows to create a virtual subclass that do not inherit from a class but are still known as such if you call e.g isinstance(). The abc module provides the abstractmethod
decorator which we use here. See: https://docs.python.org/3/library/abc.html#module-abc

Reply:
> shouldn't the FileSystemBase class be the one that inherits from ABCMeta? 

That I don't understand. Why should the base class to filesystems be abstract ?

Reply:
> Probably we should have named those classes as FileSystemFactory 

Also don't understand this. If FileSystem makes it clear by type and being abstract it is in my opinion a nice read to just say ```FileSystem.new(name="ext3" ...)``` and get a new instance of an object that handles ext3. Please note you can't create  a FileSystem instance as the constructor is abstract, you have to use the new() method which indicates it creates something new for me

Reply:
> for instance, FileSystemExt4 would be an instance of FileSystem

That would be totally confusing to me. Any specific filesystem class inherits from a common base. The factory around all filesystems is coming from FileSystem which just says "you are dealing with the filesystem api". If I would know about the api I would then lookup the base class, if I'm interested in a particular filesystem I would look up the implementation of that class

Reply:
Please look at this PR as if you would be somebody that wants to use kiwi as an api to your own program.

Reply:
> The abc module provides the abstractmethod decorator which we use here. See: https://docs.python.org/3/library/abc.html#module-abc. If you don't inherit from ABCMeta you can't make the constructor abstract.

Ok that explains my confusion, I thought that with the inheritance you added here was about trying to achieve something like `assert isinstance(mySpecifcFSObj, FileSystem)` so whatever `FileSystem.new()` was returning was still a FileSystem instance. 

Initially I though the main purpose was to fix something like `assert isinstance(FileSystem('brtfs', ...), FileSystem)` which currently fails in our code or force that any specific FileSystemBase implementation to include all the abstract methods defined in the base class, so anyone building a new filesystem class would be forced to implement them.

So just to make it explicit, the only motivation to inherit the ABCMeta class in `FileSystem` class is to make the `__init__` abstract which in turn prevents any initialization of this class, this is the case right? Sorry I did not consider that my brain tried to find to other explanations. Then is all clear to me.

Reply:
> That I don't understand. Why should the base class to filesystems be abstract ?

Just to force that any specific class implements all the expected methods, this way anyone inheriting from it would be certain to make a proper class.

Reply:
>So just to make it explicit, the only motivation to inherit the ABCMeta class in FileSystem class is to make the __init__ abstract which in turn prevents any initialization of this class, this is the case right?

right